lib/avrora/object_container_file.ex

```diff
@@ -0,0 +1,51 @@
+defmodule Avrora.ObjectContainerFile do
+  @moduledoc """
+  Function to extract schema from elavro tuple
+  """
+  alias Avrora.Config
+  @object_container_full_name_pos 6
+
+  @doc """
+  Extract schema from an OCF message
+  ## Examples
+      iex> test = <<79, 98, 106, 1, 3, 204, 2, 20, 97, 118, 114, 111, 46, 99, 111, 100, 101, 99, 8, 110, 117,
+      ...> 108, 108, 22, 97, 118, 114, 111, 46, 115, 99, 104, 101, 109, 97, 144, 2, 123, 34, 110, 97,
+      ...> 109, 101, 115, 112, 97, 99, 101, 34, 58, 34, 105, 111, 46, 99, 111, 110, 102, 108, 117, 101,
+      ...> 110, 116, 34, 44, 34, 110, 97, 109, 101, 34, 58, 34, 80, 97, 121, 109, 101, 110, 116, 34,
+      ...> 44, 34, 116, 121, 112, 101, 34, 58, 34, 114, 101, 99, 111, 114, 100, 34, 44, 34, 102, 105,
+      ...> 101, 108, 100, 115, 34, 58, 91, 123, 34, 110, 97, 109, 101, 34, 58, 34, 105, 100, 34, 44,
+      ...> 34, 116, 121, 112, 101, 34, 58, 34, 115, 116, 114, 105, 110, 103, 34, 125, 44, 123, 34, 110,
+      ...> 97, 109, 101, 34, 58, 34, 97, 109, 111, 117, 110, 116, 34, 44, 34, 116, 121, 112, 101, 34,
+      ...> 58, 34, 100, 111, 117, 98, 108, 101, 34, 125, 93, 125, 0, 236, 47, 96, 164, 206, 59, 152,
+      ...> 115, 80, 243, 64, 50, 180, 153, 105, 34, 2, 90, 72, 48, 48, 48, 48, 48, 48, 48, 48, 45, 48,
+      ...> 48, 48, 48, 45, 48, 48, 48, 48, 45, 48, 48, 48, 48, 45, 48, 48, 48, 48, 48, 48, 48, 48, 48,
+      ...> 48, 48, 48, 123, 20, 174, 71, 225, 250, 47, 64, 236, 47, 96, 164, 206, 59, 152, 115, 80,
+      ...> 243, 64, 50, 180, 153, 105, 34>>
+      iex> Avrora.ObjectContainerFile.extract_schema(test)
+      {:ok,
+        %Avrora.Schema{
+          full_name: "io.confluent.Payment",
+          id: nil,
+          json: nil,
+          lookup_table: nil,
+          version: nil
+        }
+      }
+  """
+
+  @spec extract_schema(binary()) :: {:ok, Avrora.Schema.t()}
+  def extract_schema(payload) when is_binary(payload) do
+    {_, schema_info, _} = :avro_ocf.decode_binary(payload)
+    full_name = elem(schema_info, @object_container_full_name_pos)
+    with {:ok, nil} <- memory_storage().get(full_name),
+          {:ok, schema} <- Avrora.Schema.parse(schema_info) do
+            memory_storage().put(full_name, schema)
+            {:ok, schema}
+    else
+      {:ok, schema} -> {:ok, schema}
+      {:error, reason} -> {:error, reason}
+    end
+  end
```

Comment:
Let's do it together 🤝 

First of all the `decode_binary` from `erlavro` gives us 100% of what we need to build the schema, check this out

```elixir
payload = <<79, 98, 106, 1, 3, 204, 2, 20, 97, 118, 114, 111, 46, 99, 111, 100, 101, 99, 8, 110, 117,
108, 108, 22, 97, 118, 114, 111, 46, 115, 99, 104, 101, 109, 97, 144, 2, 123, 34, 110, 97,
109, 101, 115, 112, 97, 99, 101, 34, 58, 34, 105, 111, 46, 99, 111, 110, 102, 108, 117, 101,
110, 116, 34, 44, 34, 110, 97, 109, 101, 34, 58, 34, 80, 97, 121, 109, 101, 110, 116, 34,
44, 34, 116, 121, 112, 101, 34, 58, 34, 114, 101, 99, 111, 114, 100, 34, 44, 34, 102, 105,
101, 108, 100, 115, 34, 58, 91, 123, 34, 110, 97, 109, 101, 34, 58, 34, 105, 100, 34, 44,
34, 116, 121, 112, 101, 34, 58, 34, 115, 116, 114, 105, 110, 103, 34, 125, 44, 123, 34, 110,
97, 109, 101, 34, 58, 34, 97, 109, 111, 117, 110, 116, 34, 44, 34, 116, 121, 112, 101, 34,
58, 34, 100, 111, 117, 98, 108, 101, 34, 125, 93, 125, 0, 236, 47, 96, 164, 206, 59, 152,
115, 80, 243, 64, 50, 180, 153, 105, 34, 2, 90, 72, 48, 48, 48, 48, 48, 48, 48, 48, 45, 48,
48, 48, 48, 45, 48, 48, 48, 48, 45, 48, 48, 48, 48, 45, 48, 48, 48, 48, 48, 48, 48, 48, 48,
48, 48, 48, 123, 20, 174, 71, 225, 250, 47, 64, 236, 47, 96, 164, 206, 59, 152, 115, 80,
243, 64, 50, 180, 153, 105, 34>>
:avro_ocf.decode_binary(payload)

{{:header, <<79, 98, 106, 1>>,
  [
    {"avro.codec", "null"},
    {"avro.schema",
     "{\"namespace\":\"io.confluent\",\"name\":\"Payment\",\"type\":\"record\",\"fields\":[{\"name\":\"id\",\"type\":\"string\"},{\"name\":\"amount\",\"type\":\"double\"}]}"}
  ],
  <<236, 47, 96, 164, 206, 59, 152, 115, 80, 243, 64, 50, 180, 153, 105, 34>>},
 {:avro_record_type, "Payment", "io.confluent", "", [],
  [
    {:avro_record_field, "id", "", {:avro_primitive_type, "string", []},
     :undefined, :ascending, []},
    {:avro_record_field, "amount", "", {:avro_primitive_type, "double", []},
     :undefined, :ascending, []}
  ], "io.confluent.Payment", []},
 [[{"id", "00000000-0000-0000-0000-000000000000"}, {"amount", 15.99}]]}
```

So what do we have in the output:

1. We have prepared JSON format of the schema
2. We have the `erlavro` structures for the schema (needed for the lookup table)
3. We have full name

Let's try to make use of all of them and apply some Elixir-way to write `with` condition

```suggestion
  def extract_schema(payload) when is_binary(payload) do
    with {:ok, {headers, erlavro = {_, _, _, _, _, _, full_name, _}, _}} <- do_decode(payoad),
         {:ok, nil} <- memory_storage().get(full_name),
         {:ok, json} <- extract_json(headers),
         {:ok, schema} <- Schema.new() do
      lookup_table = :avro_schema_store.add_type(erlavro, Schema.lookup_table())
      schema = %{schema | full_name: full_name, lookup_table: lookup_table, json: json}

      memory_storage().put(schema.full_name, schema)
    end
  end

  defp extract_json(headers) do
    with {_, _, meta, _} <- headers,
         {"avro.schema", json} <- Enum.find(meta, fn {key, _} -> key == "avro.schema" end) do
      {:ok, json}
    end
  end

  defp do_decode(payload) do
    {:ok, :avro_ocf.decode_binary(payload)}
  rescue
    error -> {:error, error}
  end
```

It will require to introduce `Avrora.Schema.new()` (I also thought about `Avrora.Schema.blank()`) to give us an empty schema we can fill in.

The type spec of the method should be corrected of course and take a look on `with` statement, you don't need to handle `else` it will return it by itself

Reply:
Ok, so we create an empty Schema and fill it with the json, full_name and the reference table, I was missing this part `:avro_schema_store.add_type(erlavro, Schema.lookup_table())`
Good to know about `with` statement 😮